[1단계] 문자열 덧셈 계산기

RoadtoS7 opened the pull request:
안녕하세요 리뷰어님 
문자열 덧셈 계산기를 구현했습니다!
구현 하면서 고려한 사항은 [README](https://github.com/RoadtoS7/swift-lotto/blob/5bcdb91fc6fbc901faf2c96f69b620050e7d893b/README.md) 파일에 적어놓았습니다.
피드백 부탁드립니다 🤗

RoadtoS7 committed:
feat: 문자열을 특정 문자를 기준으로 나누는 Spliter와 구분자 역할을 하는 Splitter, Splitters 구현

RoadtoS7 committed:
docs: README에 구현할 프로그래밍 내용 수정

RoadtoS7 committed:
feat: Operator 프로토콜과 덧셈을 수행하는 Plus 프로토콜 구현

RoadtoS7 committed:
feat: NumberGenerator에서 숫자를 담는 문자을 정수타입으로 변환한다

PositiveNumberGenerator는 음의정수나 숫자이외의 값을 받으면 에러를 throw한다
PositiveNumberGenerator는 빈문자열이나 nil을 받으면 0을 반환한다
PositiveNumberGenerator는 0이상의 정수를 받으면 해당 정수를 갖는 Int타입을 반환한다

RoadtoS7 committed:
refactor: PositiveNumberGenerator에서 nil,empty string이 들어온 경우

guard let을 이용하여 한번에 처리되도록 리팩토링

RoadtoS7 committed:
fix: PositiveNumberGenerator가 NumberGenerator 프로토콜을 따르도록 수정

nylah-j committed:
feat: NumberGenerator와 Splitter를 이용하여 문자를 파싱하는 InputParser 구현

InputParser는 nil, empty string이 들어왔을 때 반환 값 처리를 담당한다
NumberGenerator는 non optional값을 받아서 숫자로 반환한다 그 과정에서 발생할 수 있는 것들을 처리한다
Splitter는 nil이나 빈문자를 받으면 nil을 반환한다

RoadtoS7 committed:
feat: 주어진 operand와 operator를 순서대로 연결해서 계산하는 Calculator 구현

RoadtoS7 committed:
feat: 사용자듸 입력을 받는 InputView 구현

RoadtoS7 committed:
feat: 결과를 출력하는 OutputView 구현

RoadtoS7 committed:
feat: InputView, OutputView를 LottoTest target에 추가

RoadtoS7 committed:
feat: main 로직 구현

RoadtoS7 committed:
refactor: PositiveNumberGenerator에서 0을 zero 상수에 담아 사용하도록 수정

RoadtoS7 committed:
refactor: InputParser와 PositiveNumberGeneator의 명확한 역할 분리

InputParser는 사용자가 nil이나 빈 문자열을 입력했을 때 그 입력을 치환하는 역할을 한다
PositiveNumberGenerator는 사용자의 입력을 양의 정수 Int로 형변환하는 역할을 한다

RoadtoS7 committed:
refactor: 사용하지 않는 LottoTest파일 제거

wonhee009 commented on Lotto/Lotto/main.swift:
```diff
@@ -6,5 +6,28 @@
 
-print("Hello, World!")
+do {
+    let input = InputView.input()
```
`inputString`은 어떨까요?
`input()`만 있어서 어떤걸 반환하는지 해당 변수가 어떨걸 의미하는지 명확하게 들어나지 않는거 같습니다:)

RoadtoS7 replied:
피드백 감사합니다! 아래 피드백과 함께 [7c9120f](https://github.com/next-step/swift-lotto/pull/16/commits/7c9120f205718fad96527f3c0353b3507e896990) 커밋에서 반영했습니다!~

wonhee009 commented on Lotto/Lotto/Separators.swift:
```diff
@@ -0,0 +1,35 @@
+
+import Foundation
+
+protocol Separator {
+    var value: String { get }
+}
+
+struct Comma: Separator {
+    let value = ","
+}
+
+struct Colon: Separator {
+    let value = ":"
+}
```
`Separator`를 protocol로 정의한 이유가 있을까요?
현재는 String을 구분하는 문자가 ,와 :으로 두개이지만 /나 ! 등이 구분 문자로 추가된다면 struct가 무한정으로 늘어날 것 같습니다.
이 부분도 어느 정도 스타일 차이라고 생각해서 의견이 궁금합니다 :)

wonhee009 replied:
`Splitter`생성과 `SplitterTests` 세팅에서 `separators` array를 생성해서 넣어주고 있는데요.
이후 /나 ! 등이 구분 문자로 추가된다면 `Splitter`생성부와 `SplitterTest` 세팅부에 `separators` array를 수정해줘야할거 같습니다.
또한, 구분 문자가 [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]일 때 6이 구분 문자에서 빠지게 되는 경우, `Splitter`생성부와 `SplitterTest` 세팅부에서 6만 삭제하고 확인해야할거 같습니다.
이런 과정에서 에러가 생길 가능성과 생성부와 세팅부 두 군데의 관리 포인트가 있다고 생각합니다 :)

RoadtoS7 replied:
구분자가 수정되었을 때 테스트코드와 프로덕션 코드에 제대로 반영되지 않을 문제점이 있겠군요!!😲
새로운 관점을 배워갑니다 피드백 감사합니다!!

> Separator를 protocol로 정의한 이유가 있을까요?  

- 단일 책임 원칙에 따라서 구분자도 각 구조체의 책임으로 분리해놓으면 변경에 유연하게 대처할 수 있지 않을까..? 라고 생각했었습니다!
- 그런데 원희님께서 주신 피드백을 통해, 이렇게 구현하면 구분자의 생성부와 세팅부가 분리됨으로써  에러가 발생할 가능성을 높인다는 것을 알게 되었습니다! 
- _이 문제를 해결하기 위해서  Separator 프로토콜과 구현체를 모두 없애고, `Splitter` 에서 상수 프로퍼티로 구분자를 갖도록 구현했습니다!_
- 혹시 이 방법에서도 발생할 수 있는 문제점이 있을지 피드백 부탁드립니다!

wonhee009 replied:
현재 방법도 좋습니다 👍 
다만, 상수 프로퍼티로 구분자를 관리하기보다는 enum, CaseIterable을 통해 관리하는게 나중에 어떤 구분자들이 있는지 확인 및 관리가 쉽지 않을까 생각이 듭니다.
해당 코멘트는 여기서 마무리하고 다음 스텝에서 고려해주세요 :)

wonhee009 commented on Lotto/Lotto/Separators.swift:
```diff
@@ -0,0 +1,35 @@
+}
+
+struct Separators {
+    let value: [Separator]
```
`private`일 필요는 없나용?.?

wonhee009 commented on Lotto/Lotto/Splitter.swift:
```diff
@@ -0,0 +1,18 @@
+import Foundation
+
+struct Splitter {
+    private static let emptyString = ""
```
static으로 정의한 이유가 있으실까요?
private으로 외부에 닫혀있는 상황인데 static일 이유가 있나 해서용

RoadtoS7 replied:
다른 프로그래밍 언어 컨벤션에서 상수는 항상 static으로 선언하는 컨벤션이 있는데, 그것에 너무 익숙한 나머지 여기서도 습관적으로 static으로 선언했습니다 😨
[6743943](https://github.com/next-step/swift-lotto/pull/16/commits/6743943753639bb41d00ef06f852f15a2dea3970) 커밋에서  static을 제거하고 지역 상수로 수정했습니다! 
피드백 감사합니다!!

wonhee009 commented on Lotto/Lotto/View/OutputView.swift:
```diff
@@ -0,0 +1,14 @@
+import Foundation
+
+struct OutputView {
+    static func showResult(_ result: Int) {
```
`print(result: Int)`는 어떨까요?
show보다는 print가 해당 역할에 더 맞는거 같습니다.
또한, `printResult(_ result: Int)`의 result 중복도 피할 수 있을거 같구요 :)

RoadtoS7 replied:
오 원희님께서 피드백 주신 방향으로 수정해보니 더 코드가 깔끔해졌습니다! 
피드백 감사합니다!~
[87c1ad6](https://github.com/next-step/swift-lotto/pull/16/commits/87c1ad6389d8ed6b2d779a6055fe94cb63e262c0) 커밋에서 반영했습니다!

wonhee009 commented on Lotto/Lotto/View/InputView.swift:
```diff
@@ -0,0 +1,15 @@
+import Foundation
+
+struct InputView {
+    static func input() -> String? {
```
`readInput()`은 어떨까요?
`input()`만으로는 해당 메서드의 역할을 명확하게 알기 어려울거 같습니다 :)

RoadtoS7 replied:
피드백 주신 것이 더 메서드의 의미 파악이 잘 되는 것 같습니다! 피드백 감사합니다!! 
[7c9120f](https://github.com/next-step/swift-lotto/pull/16/commits/7c9120f205718fad96527f3c0353b3507e896990) 커밋에서 반영했습니다!~

wonhee009 commented on README.md:
```diff
@@ -2,3 +2,36 @@
+2. 음수가 가능해질 수 있다. 
+- 양수, 음수, double, 소숫점 계산도 지원한다면 어떻게 해야하는가?
+- 오른쪽으로 갈 수록 범위가 커지고, 왼쪽의 것을 포함한다. (UInt, Int, Float, Double)  
+- UIntGenerator, IntGEnerator, FloatGenerator, DoubleGenerator 이런식으로 각각의 Generator를 만들면 어떨까?    
```
```suggestion
- UIntGenerator, IntGenerator, FloatGenerator, DoubleGenerator 이런식으로 각각의 Generator를 만들면 어떨까?    
```

RoadtoS7 replied:
엇 오타네요..! 😅 피드백 감사합니다! 
[2e6e7ca](https://github.com/next-step/swift-lotto/pull/16/commits/2e6e7ca189437b8d67146e40ea7f6bb4d640f80e) 커밋에서 반영했습니다!

wonhee009 commented on Lotto/LottoTest/SplitterTest.swift:
```diff
@@ -0,0 +1,30 @@
+class SplitterTest: XCTestCase {
+    var splitter: Splitter!
+    
+    override func setUpWithError() throws {
+        let separators = Separators(value: [Comma(), Colon()])
+        splitter = Splitter(separaters: separators)
+    }
```
`tearDownWithError`도 정의해 주는건 어떨까요?

RoadtoS7 replied:
```Swift
override func tearDownWithError() throws {
        splitter = nil
}
```

이런식으로 `tearDownWithError`를 작성했는데 테스트 코드에 익숙치 않아서 이렇게 작성하는 것이 맞는지 궁금합니다..!

wonhee009 replied:
네네 맞습니다 :)
현재는 해당 부분이 없어도 되지만 이후 복잡한 테스트 메서드들을 작성한다면 프로퍼티들을 초기화해주는게 중요하기도 합니다!

wonhee009 commented on Lotto/Lotto/InputParser.swift:
```diff
@@ -0,0 +1,26 @@
+    
+    func parse(input: String?) throws -> [Int] {
+        guard let input = input,
+              input.isEmpty == false else {
```
```swift
extension String {
   var isNotEmpty: Bool {
      return !self.isEmpty
   }
}
```
이렇게 extension으로 정의해두고 `input.isNotEmpty`로 사용해도 좋을거 같습니다 :)

RoadtoS7 replied:
오 이렇게 사용하는 방법도 있군요! 피드백 감사합니다!! 
[ac2fde7](https://github.com/next-step/swift-lotto/pull/16/commits/ac2fde7975de9bddbcbc425ed915977b4ce1ebef) 커밋에서 반영했습니다!

wonhee009 commented on Lotto/Lotto/NumberGenerator.swift:
```diff
@@ -0,0 +1,42 @@
+    }
+    
+    static func from(_ input: String) throws -> Int {
+        guard let number = Int(input) else {
+            throw Self.Error.isNotNumber(input)
+        }
+        
+        if number < zero {
+            throw Self.Error.isNegativeNumber(number)
+        }
```
```suggestion
        guard let number = Int(input) else {
            throw Error.isNotNumber(input)
        }
        
        if number < zero {
            throw Error.isNegativeNumber(number)
        }
```

RoadtoS7 replied:
객체 지향 생활체조에서 '한 줄에 점을 하나만 사용합니다'에 어긋나는 내용이라서 피드백 주신 것으로 이해했는데 제가 제대로 이해한 것이 맞는지 궁금합니다! 

[063b13e](https://github.com/next-step/swift-lotto/pull/16/commits/063b13efadc6f484ead70e073cf129c90c673734) 커밋에서 반영했습니다!

wonhee009 replied:
객체 지향 생활체조 <- 이런 부분도 있지만, `Self`나 `self`를 코드에 사용할때에도 이유가 있어야 한다고 생각합니다.
지금 코드도 `Self`가 있으나 없으나 굴러가는데에는 문제가 없는데요.
`Self`가 써져 있다면 타입으로 어떤 일을 하나?라는 암시를 줄 수 있다고 생각합니다.
(self의 경우도 마찬가지입니다.)

wonhee009 reviewed (changes requested):
수고하셨습니다 🎉 
코멘트 남겨두었으니 확인 부탁드려요 🙏

RoadtoS7 committed:
refactor: OutputView에서 print 함수이름 변경

OutputView에서 결과를 '출력'하는 기능이라는 점에서 print를 사용하는 것이 더 가독성이 좋다
OutputView.show -> OutputView.print로 메서드 이름을 변경함

RoadtoS7 committed:
refactor: InputView에서 입력을 받는 메서드 이름 변경

InputView에서 input보다 readInput이 입력을 받아드린다는 점에서 더 구체적인 메서드명이다

RoadtoS7 committed:
doc: 오타 수정

RoadtoS7 committed:
refactor: PositiveNumberGenerator에서 Error참조시 Self제거

RoadtoS7 committed:
fix: main에서 InputView, OutputView 변경사항 미반영으로 인한 오류 해결

InputView와 OutputView의 변경사항 반영하여 해결

RoadtoS7 committed:
refactor: Splitter에서 emptyString을 지역상수로 변경

Splitter 내부에서 private으로 사용되기 때문에 지역 상수로 변경함

RoadtoS7 committed:
refactor: SplitterTest에서 tearDownWithError 작성

setUpWithError에서 생성한 splitter 인스턴스를 tearDownWithError에서 해제한다

RoadtoS7 committed:
refactor: InputParserTest에서 tearDownWithError 작성

setUpWithError()에서 생성한 InputParser 인스턴스를 tearDownWithError 에서 해제한다

RoadtoS7 committed:
refactor: String에 isNotEmpty 프로퍼티 구현 및 InputParser에 적용

RoadtoS7 committed:
fix: test 빌드에서 String extension 사용할 수 있도록 수정

RoadtoS7 committed:
refactor: Separator 프로토콜 제거, Splitter에서 구분자 프로퍼티로 가짐

- 문제점
구분자를 생성하는 곳과 설정하는 곳이 분리되어있기 때문에,
구분자가 변경되었을 때 생성부분은 변경했으나 설정 부분을 변경하지 않았을 경우 에러가 발생할 위험이 있었다.

- 해결 방안
따라서 구분자를 변경했을 때 구분자를 사용하는 곳에서도 이 변경이 바로 반영되도록 함으로써 이를 해결하고자 했다.
이를 위해서 구분자를 사용하는 곳인 Splitter에서 구분자를 상수 프로퍼티로 갖고 이용하도록 했다.

wonhee009 commented on Lotto/Lotto/NumberGenerator.swift:
```diff
@@ -0,0 +1,42 @@
+}
+
+struct PositiveNumberGenerator: NumberGenerator {
+    private static let zero = 0
```
이 부분에서도 `static`이 필요 한가요? 👀

RoadtoS7 replied:
아니요..! 이 static이 여기도 있었군요! 😱
피드백 감사합니다!
[d51b331](https://github.com/next-step/swift-lotto/pull/16/commits/d51b3316ca2e2d22b4cef53f9734e3f50f70b3c5) 커밋에서 반영했습니다!

RoadtoS7 committed:
refactor: PositiveNumberGenerator에서 zero를 지역상수 프로퍼티로 변경

PositiveNumberGenerator에서 0을 나타내는 zero 상수가 PositiveNumberGenerator에서만 사용되기 때문에 static으로 존재할 필요가 없다
따라서 지역 상수 프로퍼티로 수정한다

wonhee009 reviewed (approved):
수고하셨습니다 :)
커밋이 잘 나눠져 있고, 자세하게 적혀있어 확인하기 편했습니다.
다음 스텝에서 봐요 😄
